crates/stages-api/src/pipeline/mod.rs

```diff
@@ -129,17 +129,33 @@ where
     /// Consume the pipeline and run it until it reaches the provided tip, if set. Return the
     /// pipeline and its result as a future.
     #[track_caller]
-    pub fn run_as_fut(mut self, tip: Option<B256>) -> PipelineFut<DB> {
+    pub fn run_as_fut(mut self, target: Option<PipelineTarget>) -> PipelineFut<DB> {
         // TODO: fix this in a follow up PR. ideally, consensus engine would be responsible for
         // updating metrics.
         let _ = self.register_metrics(); // ignore error
         Box::pin(async move {
             // NOTE: the tip should only be None if we are in continuous sync mode.
-            if let Some(tip) = tip {
-                self.set_tip(tip);
+            if let Some(target) = target {
+                match target {
+                    PipelineTarget::Sync(tip) => self.set_tip(tip),
+                    PipelineTarget::Unwind(target) => {
+                        if let Err(err) = self.produce_static_files() {
+                            return (self, Err(err.into()))
+                        }
+                        if let Err(err) = self.unwind(target, None) {
+                            return (self, Err(err))
+                        }
+                        self.progress.update(target);
+
+                        // TODO: if report ControlFlow::Unwind we will issue a warning of bad block
+                        // detected. Do we want?
+                        return (self, Ok(ControlFlow::Continue { block_number: target }))
```

Comment:
yeah we shouldn't report Unwind if it's not due to an invalid block. Maybe we should make unwind into an enum, for "invalid block unwinds" and other unwinds.